google-cloud-firestore/src/main/java/com/google/cloud/firestore/AggregateQuery.java

```diff
@@ -112,52 +143,156 @@ private final class AggregateQueryResponseDeliverer {
       this.startTimeNanos = startTimeNanos;
     }
 
-    ApiFuture<AggregateQuerySnapshot> getFuture() {
+    @Nullable
+    ByteString getTransactionId() {
+      return transactionId;
+    }
+
+    @Nullable
+    com.google.protobuf.Timestamp getReadTime() {
+      return readTime;
+    }
+
+    long getStartTimeNanos() {
+      return startTimeNanos;
+    }
+
+    @Nullable
+    ExplainOptions getExplainOptions() {
+      return null;
+    }
+
+    boolean futureNotCompleted() {
+      return futureCompleted.compareAndSet(false, true);
+    }
+
+    ApiFuture<T> getFuture() {
       return future;
     }
 
-    void deliverResult(@Nonnull Map<String, Value> data, Timestamp readTime) {
-      if (isFutureCompleted.compareAndSet(false, true)) {
-        Map<String, Value> mappedData = new HashMap<>();
-        data.forEach((serverAlias, value) -> mappedData.put(aliasMap.get(serverAlias), value));
-        future.set(new AggregateQuerySnapshot(AggregateQuery.this, readTime, mappedData));
-      }
+    void setFuture(T value) {
+      future.set(value);
     }
 
     void deliverError(Throwable throwable) {
-      if (isFutureCompleted.compareAndSet(false, true)) {
+      if (futureNotCompleted()) {
         future.setException(throwable);
       }
     }
+
+    abstract void deliverResult(
+        @Nullable Map<String, Value> serverData,
+        Timestamp readTime,
+        @Nullable ExplainMetrics metrics);
   }
 
-  private final class AggregateQueryResponseObserver
-      implements ResponseObserver<RunAggregationQueryResponse> {
+  private class AggregateQueryResponseDeliverer extends ResponseDeliverer<AggregateQuerySnapshot> {
+    AggregateQueryResponseDeliverer(
+        @Nullable ByteString transactionId,
+        @Nullable com.google.protobuf.Timestamp readTime,
+        long startTimeNanos) {
+      super(transactionId, readTime, startTimeNanos);
+    }
 
-    private final AggregateQueryResponseDeliverer responseDeliverer;
-    private StreamController streamController;
+    @Override
+    void deliverResult(
+        @Nullable Map<String, Value> serverData,
+        Timestamp readTime,
+        @Nullable ExplainMetrics metrics) {
+      if (futureNotCompleted()) {
+        if (serverData == null) {
+          throw new RuntimeException("Did not receive any aggregate query results.");
+        }
+        setFuture(
```

Comment:
The `setFuture` has built in control to make sure it is only called once. From docs:

> When a call to this method returns, the Future is guaranteed to be done only if the call was accepted (in which case it returns true).

The `AtomicBoolean` is not required.

Reply:
Done